src/persistQueryClient/index.ts

```diff
@@ -84,8 +90,53 @@ export async function persistQueryClient({
       )
       persister.removeClient()
     }
+  }
+}
+
+/**
+ * Persists data from the QueryCache
+ *  - data dehydrated using dehydrateOptions
+ *  - data is persisted using persister.persistClient
+ */
+export async function persistQueryClientSave({
+  queryClient,
+  persister,
+  buster = '',
+  dehydrateOptions,
+}: PersistedQueryClientSaveOptions) {
+  if (typeof window !== 'undefined') {
+    const persistClient: PersistedClient = {
+      buster,
+      timestamp: Date.now(),
+      clientState: dehydrate(queryClient, dehydrateOptions),
+    }
+
+    await persister.persistClient(persistClient)
+  }
+}
+
+/**
+ * Subscribe to QueryCache updates (for persisting)
+ * @returns an unsubscribe function (to discontinue monitoring)
+ */
+export function persistQueryClientSubscribe(
+  props: PersistedQueryClientSaveOptions
+) {
+  return props.queryClient.getQueryCache().subscribe(() => {
+    persistQueryClientSave(props)
+  })
+}
+
+/**
+ * Restores persisted data to QueryCache and persists further changes.
+ * (Retained for backwards compatibility)
+ */
+export async function persistQueryClient(props: PersistQueryClientOptions) {
+  if (typeof window !== 'undefined') {
+    // Attempt restore
+    persistQueryClientRestore(props)
```

Comment:
I think we need to `await persistQueryClientRestore(props)` before we trigger the subscription, do we not?

Reply:
My reading of the existing version is that it does not await:  

https://github.com/tannerlinsley/react-query/blob/d7faf86005052654451dd7d5b3665bf6c971b66e/src/persistQueryClient-experimental/index.ts#L63-L89

However, I can add it if you like?

Reply:
> the try is synchronous?

I don't think it is. The whole function is an async function, and in line 65 we `await persister.restoreClient() `. If we move that to a separate function, we have to await it to behave the same as before.

Reply:
https://github.com/tannerlinsley/react-query/pull/3141/commits/df7686c11dfc58bc8989b03bb3f9565040bdabd2